With NET 5 RC1 x86 SDK installed on x64 OS, run Unit test failed with error: Testhost process exited with error: A fatal error occurred, the required library hostfxr.dll could not be found...

Machine OS: win10-x64
**Repro Steps**
Install NET 5 RC1 x86 SDK from https://github.com/dotnet/installer on x64 OS
dotnet new xunint/mstest/nunit in CLI
dotnet build and then dotnet test

**Expected result:**
Unit test run successfully

**Actual result:**
run failed with below error:
![image](https://user-images.githubusercontent.com/65637796/91423723-7e94ca80-e88b-11ea-80df-cf8ec8273cd0.png)
Testhost process exited with error: A fatal error occurred, the required library hostfxr.dll could not be found.
If this is a self-contained application, that library should exist in [C:\Users\v-danche\today\hahah\bin\Debug\net5.0\].
If this is a framework-dependent application, install the runtime in the default location [C:\Program Files\dotnet] or use the DOTNET_ROOT environment variable to specify the runtime location.
. Please check the diagnostic logs for more information.

Test Run Aborted.
C:\Program Files (x86)\dotnet\sdk\5.0.100-rc.1.20427.1\Microsoft.TestPlatform.targets(32,5): error MSB4181: The "Microsoft.TestPlatform.Build.Tasks.VSTestTask" task returned false but did not log an error. [C:\Users\v-danche\today\hahah\hahah.csproj]

**Note:
It doesn't repro on x64 SDK.**
It only repro on NET 5 RC1 x86.

Dotnet --info:
.NET SDK (reflecting any global.json):
 Version:   5.0.100-rc.1.20427.1
 Commit:    dd2e854c91

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.18362
 OS Platform: Windows
 RID:         win10-x86
 Base Path:   C:\Program Files (x86)\dotnet\sdk\5.0.100-rc.1.20427.1\

Host (useful for support):
  Version: 5.0.0-rc.1.20426.10
  Commit:  bfc00e4e13

.NET SDKs installed:
  5.0.100-rc.1.20427.1 [C:\Program Files (x86)\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 5.0.0-rc.1.20426.3 [C:\Program Files (x86)\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 5.0.0-rc.1.20426.10 [C:\Program Files (x86)\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 5.0.0-rc.1.20426.4 [C:\Program Files (x86)\dotnet\shared\Microsoft.WindowsDesktop.App]

To install additional .NET runtimes or SDKs:
  https://aka.ms/dotnet-download

@pavelhorak, I can repro this issue on with NET 5 RC1 x86 - (5.0.100-rc.1.20454.5)
C:\Users\v-danche>"%ProgramFiles(x86)%/dotnet/dotnet" --info
.NET SDK (reflecting any global.json):
 Version:   5.0.100-rc.1.20454.5
 Commit:    53a032d031

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.18362
 OS Platform: Windows
 RID:         win10-x86
 Base Path:   C:\Program Files (x86)\dotnet\sdk\5.0.100-rc.1.20454.5\

Host (useful for support):
  Version: 5.0.0-rc.1.20451.14
  Commit:  38017c3935

.NET SDKs installed:
  5.0.100-rc.1.20454.5 [C:\Program Files (x86)\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 5.0.0-rc.1.20451.17 [C:\Program Files (x86)\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 5.0.0-rc.1.20451.14 [C:\Program Files (x86)\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 5.0.0-rc.1.20452.2 [C:\Program Files (x86)\dotnet\shared\Microsoft.WindowsDesktop.App]

To install additional .NET runtimes or SDKs:
  https://aka.ms/dotnet-download

C:\Users\v-danche>cd C:\Users\v-danche\test

C:\Users\v-danche\test>"%ProgramFiles(x86)%/dotnet/dotnet" new mstest -o mymstest
The template "Unit Test Project" was created successfully.

Processing post-creation actions...
Running 'dotnet restore' on mymstest\mymstest.csproj...
  Determining projects to restore...
  Restored C:\Users\v-danche\test\mymstest\mymstest.csproj (in 360 ms).
Restore succeeded.


C:\Users\v-danche\test>cd mymstest

C:\Users\v-danche\test\mymstest>"%ProgramFiles(x86)%/dotnet/dotnet" build
Microsoft (R) Build Engine version 16.8.0-preview-20451-02+51a1071f8 for .NET
Copyright (C) Microsoft Corporation. All rights reserved.

  Determining projects to restore...
  All projects are up-to-date for restore.
  You are using a preview version of .NET. See: https://aka.ms/dotnet-core-preview
  mymstest -> C:\Users\v-danche\test\mymstest\bin\Debug\net5.0\mymstest.dll

Build succeeded.
    0 Warning(s)
    0 Error(s)

Time Elapsed 00:00:01.29

C:\Users\v-danche\test\mymstest>"%ProgramFiles(x86)%/dotnet/dotnet" test
  Determining projects to restore...
  All projects are up-to-date for restore.
  You are using a preview version of .NET. See: https://aka.ms/dotnet-core-preview
  mymstest -> C:\Users\v-danche\test\mymstest\bin\Debug\net5.0\mymstest.dll
Test run for C:\Users\v-danche\test\mymstest\bin\Debug\net5.0\mymstest.dll (.NETCoreApp,Version=v5.0)
Microsoft (R) Test Execution Command Line Tool Version 16.8.0
Copyright (c) Microsoft Corporation.  All rights reserved.

Starting test execution, please wait...
A total of 1 test files matched the specified pattern.
Testhost process exited with error: A fatal error occurred, the required library hostfxr.dll could not be found.
If this is a self-contained application, that library should exist in [C:\Users\v-danche\test\mymstest\bin\Debug\net5.0\].
If this is a framework-dependent application, install the runtime in the default location [C:\Program Files\dotnet] or use the DOTNET_ROOT environment variable to specify the runtime location.
. Please check the diagnostic logs for more information.

Test Run Aborted.
C:\Program Files (x86)\dotnet\sdk\5.0.100-rc.1.20454.5\Microsoft.TestPlatform.targets(32,5): error MSB4181: The "Microsoft.TestPlatform.Build.Tasks.VSTestTask" task returned false but did not log an error. [C:\Users\v-danche\test\mymstest\mymstest.csproj]

C:\Users\v-danche\test\mymstest>

> 1. Install 64-bit SDK (or just runtime), this will allow dotnet test to find it and use it to run tests.

I don't think this is a valid option. At least in my CI scenarios, I explicitly download and install the 32-bit SDK because I am testing my binaries when run under 32-bit mode.
Installing the 64-bit SDK and having it run the tests in 64-bit mode defeats that.

> 2. Tell dotnet test explicity which platform to use for the run by -p:Platform=x86.

This doesn't work at the solution level and you will get an error such as:
```
 C:\Repos\terrafx.interop.windows\TerraFX.Interop.Windows.sln.metaproj : error MSB4126: The specified solution configuration "Debug|x86" is invalid. Please specify a valid solution configuration using the Configuration and Platform properties (e.g. MSBuild.exe Solution.sln /p:Configuration=Debug /p:Platform="Any CPU") or leave those properties blank to use the default solution configuration. [C:\Repos\terrafx.interop.windows\TerraFX.Interop.Windows.sln]
```

> Build your DLL as x86 instead of AnyCPU.

This is the same issue as 2 and shouldn't be required, IMO. You should be able to just do something akin to `dotnet test --architecture x86` and have it run the tests as `x86`, regardless of whether you are running from a 32-bit or 64-bit SDK.
I imagine this also exists for `arm` vs `arm64` where another Windows on Windows layer exists (and where `x86` is also a valid target via the emulation layer).

> If this is a test scenario, I think it should be re-visited. At one point in time we broke this logic and got multiple reports of dotnet test incorrectly looking for x86 SDK which was not present on the system (64-bit OS with only 64-bit SDK). I don't think we ever got the report of this inverse case (64-bit OS with only 32-bit SDK, apart from you reporting it.

Why is there any guessing here at all? If `dotnet` is 32-bit, you can safely default to running 32-bit tests. Likewise if `dotnet` (or `other host`) is 64-bit, you can safely default to it running as 64-bit.
Having the ability to explicitly specify the target such as via `--architecture` then provides a fallback so you can run `64-bit` tests from a 32-bit SDK or vice-versa.

--- a/src/vstest.console/TestPlatformHelpers/TestRequestManager.cs
+++ b/src/vstest.console/TestPlatformHelpers/TestRequestManager.cs
@@ -451,7 +451,19 @@ namespace Microsoft.VisualStudio.TestPlatform.CommandLine.TestPlatformHelpers
                         || chosenFramework.Name.IndexOf("netcoreapp", StringComparison.OrdinalIgnoreCase) >= 0
                         || chosenFramework.Name.IndexOf("net5", StringComparison.OrdinalIgnoreCase) >= 0)
                     {
-                        defaultArchitecture =  Environment.Is64BitOperatingSystem ? Architecture.X64 : Architecture.X86;
+#if NETCOREAPP
+                        // We are running in vstest.console that is either started via dotnet.exe or via vstest.console.exe .NET Core 
+                        // executable. For AnyCPU dlls this should resolve 32-bit SDK when running from 32-bit dotnet process and 
+                        // 64-bit SDK when running from 64-bit dotnet process.
+                        defaultArchitecture = Environment.Is64BitProcess ? Architecture.X64 : Architecture.X86;
+#else
+                        // We are running in vstest.console.exe that was built against .NET Framework. This console prefers 32-bit
+                        // because it needs to run as 32-bit to be compatible with QTAgent. It runs as 32-bit both under VS and 
+                        // in Developer console. Set the default architecture based on the OS architecture, to find 64-bit dotnet SDK
+                        // when running AnyCPU dll on 64-bit system, and 32-bit SDK when running AnyCPU dll on 32-bit OS.
+                        // We want to find 64-bit SDK because it is more likely to be installed.
+                        defaultArchitecture = Environment.Is64BitOperatingSystem ? Architecture.X64 : Architecture.X86;
+#endif
                     }
 
                     settingsUpdated |= this.UpdatePlatform(document, navigator, sources, sourcePlatforms, defaultArchitecture, out Architecture chosenPlatform);

